legate/core/legion.py

```diff
@@ -178,49 +180,51 @@ def __init__(self, p=None, dim=None):
             self.set_point(p)
 
     @property
-    def dim(self):
+    def dim(self) -> int:
         return self.point.dim
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: int) -> Any:
         if key >= self.dim:
             raise KeyError("key cannot exceed dimensionality")
         return self.point.point_data[key]
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key: int, value: Any) -> None:
         if key >= self.dim:
             raise KeyError("key cannot exceed dimensionality")
         self.point.point_data[key] = value
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         value = hash(self.dim)
         for idx in range(self.dim):
             value = value ^ hash(self[idx])
         return value
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Point):
+            return False
```

Comment:
Is this the common practice or `type(self) == type(other)`? I have used explicit type equality in the past, because (I think) that avoids the weird situation where `superclass_instance == subclass_instance` but  `subclass_instance != superclass_instance`.

Reply:
I've definitely seen both, and sometimes even in the same project (both are present inside the Numba codeabse, e.g.) The main motivation to switch is because `mypy` is smart enough to reason about the `isinstance` check, but not smart enough to reason about the other check. We would need a casts for that case, which is not great but also not the end of the world.

However, you encouraged me to dig a bit and I think the actual solution here might simply be to return [`NotImplemented`](https://docs.python.org/3/library/constants.html#NotImplemented) instead of False.
> When a binary (or in-place) method returns NotImplemented the interpreter will try the reflected operation on the other type

That should make `sub == super` so long as `super == sub`

<details> 

```
In [1]: class sup:
   ...:     def __init__(self, a):
   ...:         self.a = a
   ...:     def __eq__(self, other):
   ...:         if not isinstance(other, sup):
   ...:             return NotImplemented
   ...:         return self.a == other.a
   ...: 

In [2]: class sub(sup):
   ...:     def __init__(self, a, b):
   ...:         super().__init__(a)
   ...:         self.b = b
   ...:     def __eq__(self, other):
   ...:         if not isinstance(other, sub):
   ...:             return NotImplemented
   ...:         return self.a == other.a and self.b == other.b
   ...: 

In [3]: P = sup(10)

In [4]: C = sub(10, 20)

In [5]: P == C
Out[5]: True

In [6]: C == P
Out[6]: True
```
</details>

I will make push this change if that seems good to you. All this said, except for the partition functors, all the classes in `legion.py` are concrete.

Reply:
Ah, and style is exactly what mypy itself recommends if you set a bad type for `other`:
```
egate/core/legion.py:573:5: note: This violates the Liskov substitution principle
legate/core/legion.py:573:5: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
legate/core/legion.py:573:5: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
legate/core/legion.py:573:5: note:     def __eq__(self, other: object) -> bool:
legate/core/legion.py:573:5: note:         if not isinstance(other, AffineTransform):
legate/core/legion.py:573:5: note:             return NotImplemented
legate/core/legion.py:573:5: note:         return <logic to compare two AffineTransform instances>

```

Reply:
https://github.com/nv-legate/legate.core/pull/158/commits/ccfd31f1aeb43bead06d26f152624f0e08ab2495